src/contracts.rs

```diff
@@ -44,6 +46,9 @@ pub trait SwapCoin {
     fn get_multisig_redeemscript(&self) -> Script;
     fn get_contract_tx(&self) -> Transaction;
     fn get_contract_redeemscript(&self) -> Script;
+    fn get_other_privkey(&self) -> Option<&SecretKey>;
+    fn get_contract_privkey(&self) -> Option<&SecretKey>;
+    fn get_hash_preimage(&self) -> Option<&[u8; 32]>;
```

Comment:
These functions should not be part of the trait SwapCoin.

Because WatchOnlySwapCoins will never have access to these, and our code ends up trying to get the hash preimage or contract privkey from a WatchOnlySwapCoin then the compiler should raise an error. I got the idea from this section https://doc.rust-lang.org/book/ch17-03-oo-design-patterns.html#encoding-states-and-behavior-as-types

My other comments right now will explain how the code can be modified to never require these three functions in the SwapCoin trait

Reply:
I removed these functions from `SwapCoin`. Added a couple other functions to get the contract type, and whether contract information is known (`other_privkey` for `IncomingSwapCoin`, `hash_preimage` for `Incoming/OutgoingSwapCoin`), and the type string.